crates/ruff_linter/src/importer/mod.rs

```diff
@@ -524,6 +531,17 @@ impl<'a> Importer<'a> {
         }
     }
 
+    /// Find the last `from __future__` import statement in the AST.
+    fn find_last_future_import(&self) -> Option<&'a Stmt> {
+        self.python_ast.iter().rev().find(|stmt| {
+            if let Stmt::ImportFrom(import_from) = stmt {
+                import_from.module.as_deref() == Some("__future__")
+            } else {
+                false
+            }
+        })
```

Comment:
Since future imports can only be at the beginning, iterating over all statements is inefficient, particularly on large files with many statements. Instead, only look through statements until you've found something that's *not* a future import.

Eg, something like:

```rust
fn find_last_future_import(&self) -> Option<&'a Stmt> {
  self.python_ast.iter().take_while(|stmt| {
    if let Stmt::ImportFrom(import_from) = stmt {
      import_from.module.as_deref() == Some("__future__")
    } else {
      false
    }
  }).last()
}
```